File: 43-multiply-strings/multiply-strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        if (num1 == "0" || num2 == "0") return "0";
        int* temp = new int[num1.length() + num2.length()]();

        reverse(num1.begin(), num1.end());
        reverse(num2.begin(), num2.end());
        for (int i = 0; i < num1.length(); i++){
            for (int j = 0; j < num2.length(); j++){
                int k = (num1[i] - '0') * (num2[j] - '0');
                temp[i + j] += k;
                temp[i + j + 1] += temp[i + j] / 10;
                temp[i + j] = temp[i + j] % 10;
            }
        }
        reverse(temp, temp + num1.length() + num2.length());



        int begin = 0;
        while (begin < (num1.length() +  num2.length()) && temp[begin] == 0)
            begin += 1;

        stringstream result;
        for (int i = begin; i < (num1.length() + num2.length()); i++)
            result << temp[i];
        return result.str();
    }
};

File: 43-multiply-strings/multiply-strings.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if "0" in [num1, num2]: return "0"
        result = [0] * (len(num1) + len(num2))
        num1 = num1[::-1]
        num2 = num2[::-1]
        for i in range(len(num1)):
            for j in range(len(num2)):
                result[i + j] += int(num1[i]) * int(num2[j])
                result[i + j + 1] += result[i + j] // 10
                result[i  + j] = result[i + j] % 10
        result.reverse()
        begin = 0
        for i in range(len(result)):
            if result[i] != 0:
                break
            begin += 1
        result = map(str, result[begin:])
        return ''.join(result)